docs/rfc/rfc-006-event-subscription.md

```diff
@@ -0,0 +1,189 @@
+  issue queries from the command line (e.g., using `curl`) or to write scripts
+  that call the RPC methods to monitor the running node.
+
+  For high-volume remote use, JSON-RPC is not such a good fit: Even leaving
+  aside the non-standard delivery protocol mentioned above, the time and memory
+  cost of encoding event data matters for the stability of the node when there
+  can be potentially hundreds of subscribers. Moreover, a subscription is
+  long-lived compared to most RPC methods, in that it may persist as long the
+  node is active.
+
+- **Mismatched security profiles.** The RPC service exports several methods
+  that should not be open to arbitrary remote callers, both for correctness
+  reasons (e.g., `remove_tx` and `broadcast_tx_*`) and for operational
+  stability reasons (e.g., `tx_search`). A node may still need to expose
+  events, however, to support UI tools.
+
+  This matters, because all the methods share the same network endpoint. While
+  it is possible to block the top-level GET and POST handlers with a proxy,
+  exposing the `/websocket` handler exposes not _only_ the event subscription
+  methods, but the rest of the service as well.
+
+### Possible Improvements
+
+There are several things we could do to improve the experience of developers
+who need to subscribe to events from the consensus node. These are not all
+mutually exclusive.
+
+1. **Split event subscription into a separate service**. Instead of exposing
+   event subscription on the same endpoint as the rest of the RPC service,
+   dedicate a separate endpoint on the node for _only_ event subscription.  The
+   rest of the RPC services (_sans_ events) would remain as-is.
+
+   This would make it easy to disable or firewall outside access to sensitive
+   RPC methods, without blocking access to event subscription (and vice versa).
+
+2. **Use a different protocol for event subscription.** There are various ways
+   we could approach this, depending how much we're willing to shake up the
+   current API. Here are sketches of a few options:
+
+   - Keep the websocket, but rework the API to be more JSON-RPC compliant,
+     perhaps by converting event delivery into notifications.  This is less
+     up-front change for existing clients, but retains all of the existing
+     implementation complexity, and doesn't contribute much toward more serious
+     performance and UX improvements later.
+
+   - Switch from websocket to plain HTTP, and rework the subscription API to
+     use a more conventional request/response pattern instead of streaming.
+     This is a little more up-front work for existing clients, but leverages
+     better library support for clients not written in Go. The protocol would
+     be more chatty, but we could mitigate that with batching.
+
+	 This has a nice incidental benefit that it would become easier to debug
+     subscriptions from the command-line, like the other RPC methods.
```

Comment:
what kind of responsibility for state management does this create for the service end of things? 

I'm also not sure why this is indented as much as it is. This is probably not intentional.

Reply:
Re: State management. The same as now: There would still be a queue of pending events for each subscriber. The only thing that changes is how the subscriber gets them, e.g., "give me the next up-to-k events in my queue" in a loop.

Re: Whitespace. Just accidental.

Reply:
I think we'd also need/want to add some sense of cancelation. Today, subscriptions get pruned/culled when the websockets time out, and we'll need some mechanism for deleting subscriptions that don't have anyone pulling events off of them.

Reply:
Yes, absolutely. This is just a high-level sketch of how the delivery protocol differs. All those other concerns would still remain, as now.

Reply:
It's worth noting, though, that there are some other interesting options besides just dropping the client: We could, for example, keep the subscriber connected but drop older messages if they don't pull fast enough, and the response can indicate "you missed M messages since your last call". At some point, a completely stalled client probably _should_ be dropped, but right now that's really our only practical option.

Reply:
yeah that's a cool feature. I think I like this one the best. :D